pkg/service/catalog.go

```diff
@@ -73,12 +73,15 @@ type service struct {
 // instantiated and passed to the NewPlan() constructor function which will
 // carry out all necessary initialization.
 type PlanProperties struct {
-	ID          string                 `json:"id"`
-	Name        string                 `json:"name"`
-	Description string                 `json:"description"`
-	Free        bool                   `json:"free"`
-	Metadata    *ServicePlanMetadata   `json:"metadata,omitempty"`
-	Extended    map[string]interface{} `json:"-"`
+	ID                    string                 `json:"id"`
+	Name                  string                 `json:"name"`
+	Description           string                 `json:"description"`
+	Free                  bool                   `json:"free"`
+	Metadata              *ServicePlanMetadata   `json:"metadata,omitempty"`
+	Extended              map[string]interface{} `json:"-"`
+	ProvisionParamsSchema *ParametersSchema      `json:"-"`
+	UpdateParamsSchema    *ParametersSchema      `json:"-"`
+	BindingParamsSchema   *ParametersSchema      `json:"-"`
```

Comment:
It is possible that each of the above three attributes would be better off as type `map[string]* Parameter` (or `map[string]ParameterSchema` if renamed as suggested in another comment)?

Reply:
I thought about that. In most cases where we want to provide a custom schema, we need to fetch the base properties. In that case, not sure it saves much. You need to fetch them because there _are_ some scenarios where we don't want to blindly apply them. The database only instances, for example, don't need the resource group and location. 

Can go either way with this, so if you'd like it changed to map[string]*ParameterSchema can make that change.

Reply:
> In most cases where we want to provide a custom schema, we need to fetch the base properties.

Rather than have module implementors have to retrieve the base properties and weave in their own, can we let them provide their own and `NewPlan(...)` can supplement those properties with the base properties?

> You need to fetch them because there are some scenarios where we don't want to blindly apply them. The database only instances, for example, don't need the resource group and location.

I _think_ we should be able to figure out in `NewPlan(...)` when we do and do not need to weave in the base properties. We would base it on whether the `ServiceProperties` express that the `Service` has a parent service...

The problem with that approach, however, is probably that `NewPlan(...)` doesn't have any knowledge of the `ServiceProperties`. I guess what we could do is actually make it the `NewService(...)` function iterate over all plans and augment their properties if and only if the `ServiceProperties` express that the service has a parent.

I think this would completely obviate the need for module implementors to ever concern themselves with a default / common set of properties.

Thoughts?

Reply:
Can do. This will be a little bit more involved, so probably check on this tomorrow :)